include/ginkgo/core/base/half.hpp

```diff
@@ -0,0 +1,671 @@
+                          result_traits::significand_bits,
+                  "SourceType has to have both lower range and precision or "
+                  "higher range and precision than ResultType");
+
+    static constexpr int significand_offset =
+        source_traits::significand_bits - result_traits::significand_bits;
+    static constexpr int exponent_offset = significand_offset;
+    static constexpr int sign_offset = source_traits::exponent_bits -
+                                       result_traits::exponent_bits +
+                                       exponent_offset;
+    static constexpr source_bits bias_change =
+        (source_traits::bias_mask >> exponent_offset) -
+        static_cast<source_bits>(result_traits::bias_mask);
+
+    static constexpr result_bits shift_significand(source_bits data) noexcept
+    {
+        return static_cast<result_bits>(
+            (data & source_traits::significand_mask) >> significand_offset);
+    }
+
+    static constexpr result_bits shift_exponent(source_bits data) noexcept
+    {
+        return static_cast<result_bits>(update_bias(
+            (data & source_traits::exponent_mask) >> exponent_offset));
+    }
+
+    static constexpr result_bits shift_sign(source_bits data) noexcept
+    {
+        return static_cast<result_bits>((data & source_traits::sign_mask) >>
+                                        sign_offset);
+    }
+
+private:
+    static constexpr source_bits update_bias(source_bits data) noexcept
+    {
+        return data <= bias_change ? typename source_traits::bits_type{}
+                                   : limit_exponent(data - bias_change);
+    }
+
+    static constexpr source_bits limit_exponent(source_bits data) noexcept
+    {
+        return data >= static_cast<source_bits>(result_traits::exponent_mask)
+                   ? static_cast<source_bits>(result_traits::exponent_mask)
+                   : data;
+    }
+};
+
+
+}  // namespace detail
+
+
+/**
+ * A class providing basic support for half precision floating point types.
+ *
+ * For now the only features are reduced storage compared to single precision
+ * and conversions from and to single precision floating point type.
+ */
+class half {
+public:
+    // create half value from the bits directly.
+    static constexpr half create_from_bits(uint16 bits) noexcept
+    {
+        half result;
+        result.data_ = bits;
+        return result;
+    }
+
+    // TODO: NVHPC (host side) may not use zero initialization for the data
+    // member by default constructor in some cases. Not sure whether it is
+    // caused by something else in jacobi or isai.
+    constexpr half() noexcept : data_(0){};
+
+    template <typename T, typename = std::enable_if_t<std::is_scalar<T>::value>>
+    half(const T val) : data_(0)
+    {
+        this->float2half(static_cast<float>(val));
+    }
+
+    constexpr half(const half& val) : data_(0) { data_ = val.data_; };
```

Comment:
does this have any advantage over
```suggestion
    constexpr half(const half& val) : data_(val.data_) { }
```

Reply:
no, it was changed step by step.
fixed it